### PR TITLE
Implement smarter markup escaping

### LINKED ISSUE #546: Format of the message color issue
There seems to be a problem with markup, I'm not sure where exactly is the problem, idea was to have separate colors for title and the body of the notification. I came up with this in the config file

`format = "<b><span foreground='#fcaf3e'>%s</span></b>\n%b"`

it works fine, however sometimes I did notice one problem with it but I didn't care enough to bother with it until now, problem is that in some cases title had the same color as the body, in this case white. Found out that if in the body of the message is the `&` symbol that would happen.

So, for example `dunstify "Testing" "Example"` works fine.
But `dunstify "Testing" "Example & More"` gives the problem described above.

Maybe it's escaping strings wrong or something like that, maybe there are more symbols like ampersand which would produce the same problem, I'm not sure, but I'm sure of this one.

Or maybe I'm doing something wrong with the format in the config file.

### PR DESCRIPTION
**Pull request by tsipinakis**
Automatically detect and escape invalid and unsupported & entities.

Fixes #546

### CODE REVIEW

File: src/markup.c [hunk 1]
```diff
@@ -242,6 +243,67 @@ char *markup_strip(char *str)
...
+        assert(*str == '&');
+
+        char *end = strchr(str, ';');
+        if (!end)
+                return false;
+
+        if (str[1] == '#') {
+                char *cur = str + 2;
+
+                for (; cur < end; cur++) {
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/markup.c]: And also, you're using `isalnum` here. I'm taking a guess: The entities are only hexadecimal. Then `isxdigit` would be a better match.

See https://manpages.debian.org/stretch/manpages-dev/isalnum.3.en.html#DESCRIPTION

File: src/markup.c [hunk 2]
```diff
@@ -242,6 +243,67 @@ char *markup_strip(char *str)
         return str;
 }
 
+/**
+ * Determine if an & character pointed to by \p str is a markup & entity or
+ * part of the text
+ *
+ * @return true if it's an entity otherwise false
+ */
+bool markup_is_entity(char *str)
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/markup.c]: But at least, you can use `const char *str`. You don't change anything here.

File: test/markup.c [hunk 3]
```diff
@@ -55,6 +55,11 @@ TEST test_markup_transform(void)
         ASSERT_STR_EQ("bar baz",            (ptr=markup_transform(g_strdup("<a href=\"asdf\">bar</a> baz"), MARKUP_FULL)));
         g_free(ptr);
 
+        ASSERT_STR_EQ("&#936;", (ptr=markup_transform(g_strdup("&#936;"), MARKUP_FULL)));
+        free(ptr);
+        ASSERT_STR_EQ("&amp;invalid;", (ptr=markup_transform(g_strdup("&invalid;"), MARKUP_FULL)));
```
Comment (bebehei) [inline test/markup.c]: Please add another valid string, for example `&gt;`

File: src/markup.c [hunk 4]
```diff
@@ -242,6 +243,67 @@ char *markup_strip(char *str)
...
+{
+        assert(str);
+        assert(*str == '&');
+
+        char *end = strchr(str, ';');
+        if (!end)
```
Comment (bebehei) [inline src/markup.c]: What about the input of `&;`?


```suggestion
        if (!end || end <= str+1)
```

File: src/markup.c [hunk 5]
```diff
@@ -242,6 +243,67 @@ char *markup_strip(char *str)
...
+
+/**
+ * Escape all unsupported and invalid &-entities in a string. If the resulting
+ * string does not fit it will be reallocated.
+ *
+ * @param str The string to be transformed
+ */
+char *markup_escape_unsupported(char *str)
+{
+        assert(str);
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply src/markup.c]: I'm looking at it with the mindset of "Is passing NULL likely, or does it indicate a bug?". In the current usage it's most certainly a bug that should be dealt with.

File: src/markup.c (hunk 4 above)
Comment (tsipinakis) [reply src/markup.c]: This is handled already by the named entity case (`g_str_has_prefix` fails for all supported tags with `&;` as the input).

**Comment by tsipinakis** [comment]
@bebehei Updated.

File: src/markup.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (bebehei) [reply src/markup.c]: Well, given in this context, that's right. The only context, which calls this function already asserts, that `str` is non-null.

But in the course of the last year, we converted almost all string to string functions to handle a `NULL` string by returning `NULL` as the string output. This matches completely GLib's behavior. Only if something else will get returned, then we [still `assert` the string in the function](https://github.com/dunst-project/dunst/blob/5eb34fae4e91500cf3aa93bfcae4b0a558e6b57d/src/utils.c#L145-L148).

**Comment by bebehei** [comment]
Testing was successful. Code looks very good! I even know a regular upcoming notification, will get displayed with proper formatting again :grin: Thank you :100:

**Comment by tsipinakis** [comment]
@bebehei Updated.

**Pull request merged**